### PR TITLE
Pypi packaging automated test

### PR DESCRIPTION
**Pull request by scottsand-db**
## What does this change?
Create an automated test so that the pypi delta package is installed and used in python tests every commit.

## Test / Confirmation 1
I changed the `version.sbt` version to be `version in ThisBuild := "1.1.0-SNAPSHOT-TEST"` (and did not commit those changes)

That version (with the `-TEST` appended to it) was then used when executing `using_with_pip.py`

```
### Executing: pip3 install /Users/scott.sandre/delta/dist/delta_spark-1.1.0_SNAPSHOT_TEST-py3-none-any.whl
Processing ./dist/delta_spark-1.1.0_SNAPSHOT_TEST-py3-none-any.whl
Collecting pyspark<3.2.0,>=3.1.0
  Using cached pyspark-3.1.2-py2.py3-none-any.whl
Requirement already satisfied: importlib-metadata>=3.10.0 in /usr/local/lib/python3.9/site-packages (from delta-spark==1.1.0-SNAPSHOT-TEST) (4.6.1)
Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.9/site-packages (from importlib-metadata>=3.10.0->delta-spark==1.1.0-SNAPSHOT-TEST) (3.5.0)
Requirement already satisfied: py4j==0.10.9 in /usr/local/lib/python3.9/site-packages (from pyspark<3.2.0,>=3.1.0->delta-spark==1.1.0-SNAPSHOT-TEST) (0.10.9)
Installing collected packages: pyspark, delta-spark
Successfully installed delta-spark-1.1.0-SNAPSHOT-TEST pyspark-3.1.2
Test command: ['python3', '/Users/scott.sandre/delta/examples/python/using_with_pip.py']
:: loading settings :: url = jar:file:/usr/local/lib/python3.9/site-packages/pyspark/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
Ivy Default Cache set to: /Users/scott.sandre/.ivy2/cache
The jars for the packages stored in: /Users/scott.sandre/.ivy2/jars
io.delta#delta-core_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-0b707f56-f34e-4a63-9008-b9090b7f024c;1.0
	confs: [default]
	found io.delta#delta-core_2.12;1.1.0-SNAPSHOT-TEST in local-m2-cache
	found org.antlr#antlr4;4.8 in local-m2-cache
	found org.antlr#antlr4-runtime;4.8 in local-m2-cache
	found org.antlr#antlr-runtime;3.5.2 in local-m2-cache
	found org.antlr#ST4;4.3 in local-m2-cache
	found org.abego.treelayout#org.abego.treelayout.core;1.0.3 in local-m2-cache
	found org.glassfish#javax.json;1.0.4 in local-m2-cache
	found com.ibm.icu#icu4j;61.1 in local-m2-cache
downloading file:/Users/scott.sandre/.m2/repository/io/delta/delta-core_2.12/1.1.0-SNAPSHOT-TEST/delta-core_2.12-1.1.0-SNAPSHOT-TEST.jar ...
	[SUCCESSFUL ] io.delta#delta-core_2.12;1.1.0-SNAPSHOT-TEST!delta-core_2.12.jar (4ms)
:: resolution report :: resolve 611ms :: artifacts dl 10ms
	:: modules in use:
	com.ibm.icu#icu4j;61.1 from local-m2-cache in [default]
	io.delta#delta-core_2.12;1.1.0-SNAPSHOT-TEST from local-m2-cache in [default]
	org.abego.treelayout#org.abego.treelayout.core;1.0.3 from local-m2-cache in [default]
	org.antlr#ST4;4.3 from local-m2-cache in [default]
	org.antlr#antlr-runtime;3.5.2 from local-m2-cache in [default]
	org.antlr#antlr4;4.8 from local-m2-cache in [default]
	org.antlr#antlr4-runtime;4.8 from local-m2-cache in [default]
	org.glassfish#javax.json;1.0.4 from local-m2-cache in [default]
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   8   |   2   |   2   |   0   ||   8   |   1   |
	---------------------------------------------------------------------
```

## Test / Confirmation 2
I added
```
// scalastyle:off println
println("pypi TEST")
// scalastyle:on println
```
to the top (constructor) of `DeltaLog.scala` (and did not commit those changes).

While running the `using_with_pip.py` tests, you can see that the `println` is executed:
```
########### Create a Parquet table ##############
########### Convert to Delta ###########
pypi TEST
########### Read table with DataFrames ###########
+---+
| id|
+---+
|  1|
|  4|
|  0|
|  2|
|  3|
+---+

########### Read table with DeltaTable ###########
+---+
| id|
+---+
|  1|
|  4|
|  0|
|  2|
|  3|
+---+

Therefore, this specific package version was used.
```

### CODE REVIEW

File: python/run-tests.py [hunk 1]
```diff
@@ -99,8 +99,60 @@ def run_python_style_checks(root_dir):
     run_cmd([os.path.join(root_dir, "dev", "lint-python")], stream_output=True)
 
 
+def run_pypi_packaging_tests(root_dir):
```
Comment (tdas) [inline python/run-tests.py:102]: add methods docs explaining what this is method is doing. its a little non-trivial

File: python/run-tests.py [hunk 2]
```diff
@@ -99,8 +99,60 @@ def run_python_style_checks(root_dir):
...
+
+    print("### Clearing Delta artifacts from ivy2 and mvn cache")
+    run_cmd(["rm", "-rf", "~/.ivy2/cache/io.delta/"], stream_output=True)
+    run_cmd(["rm", "-rf", "~/.m2/repository/io/delta/"], stream_output=True)
+
+    install_cmd = ["pip3", "install", "wheel", "twine", "setuptools", "--upgrade"]
+    print("### Executing:" + " ".join(install_cmd))
+    run_cmd(install_cmd, stream_output=True)
+
+    dist_dir = path.join(root_dir, "dist")
```
Comment (tdas) [inline python/run-tests.py:120]: pip_dist_dir or wheel_dist_dir

File: python/run-tests.py [hunk 3]
```diff
@@ -99,8 +99,60 @@ def run_python_style_checks(root_dir):
...
+
+    gen_artifacts_cmd_2 = ["python3", "setup.py", "sdist"]
+    print("### Executing: " + " ".join(gen_artifacts_cmd_2))
+    run_cmd(gen_artifacts_cmd_2, stream_output=True)
+
+    # we need, for example, 1.1.0_SNAPSHOT not 1.1.0-SNAPSHOT
+    version_formatted = version.replace("-", "_")
+    delta_whl_name = "delta_spark-" + version_formatted + "-py3-none-any.whl"
+
+    install_whl_cmd = ["pip3", "install", path.join(dist_dir, delta_whl_name)]
```
Comment (tdas) [inline python/run-tests.py:137]: does this automatically install pyspark as well??

File: python/run-tests.py [hunk 4]
```diff
@@ -99,8 +99,60 @@ def run_python_style_checks(root_dir):
...
+    print("### Executing: " + " ".join(install_whl_cmd))
+    run_cmd(install_whl_cmd, stream_output=True)
+
+    # run test python file directly with python and not with spark-submit
+    test_file = path.join(root_dir, path.join("examples", "python", "using_with_pip.py"))
```
Comment (tdas) [inline python/run-tests.py:142]: i am not sure how this is actually working. you are not really installing the maven artifact generated by current java/scala code. So this is probably running against the latest release maven artifacts. 

[This is the code](https://github.com/delta-io/delta/blob/ef76d4dc45748b2d3f7781555a740fe8954fde62/python/delta/pip_utils.py#L55) that is figuring out what version of delta to use when configuring sparksession to automatically download the delta maven artifact. my  guess is that this is still using delta 1.0 as that is the latest released version.

this is not right.. because our goal is to test the packaging with everything in the current code.  I think  you have to figure out a way to override this behavior. one way is to add an env var in the above linked code that specifies (i.e., 
 overrides importlib_metadata.version) which version to use for configuring spark session

File: python/run-tests.py (hunk 3 above)
Comment (scottsand-db) [reply python/run-tests.py:137]: Yes. It will output "Successfully installed delta-spark-1.1.0-SNAPSHOT pyspark-3.1.2"

File: python/run-tests.py [hunk 5]
```diff
@@ -40,7 +40,7 @@ def test(root_dir, package):
                    "--packages", package, test_file]
             print("Running tests in %s\n=============" % test_file)
             print("Command: %s" % str(cmd))
-            run_cmd(cmd, stream_output=True)
+            run_cmd(cmd, stream_output=True, print_cmd=False)
```
Comment (tdas) [inline python/run-tests.py:43]: why this change?

Comment (tdas) [reply python/run-tests.py:43]: why not print the cmd?

File: python/run-tests.py [hunk 6]
```diff
@@ -109,40 +121,35 @@ def run_pypi_packaging_tests(root_dir):
     # uninstall packages if they exist
     run_cmd(["pip3", "uninstall", "--yes", "delta-spark", "pyspark"], stream_output=True)
 
-    print("### Clearing Delta artifacts from ivy2 and mvn cache")
-    run_cmd(["rm", "-rf", "~/.ivy2/cache/io.delta/"], stream_output=True)
-    run_cmd(["rm", "-rf", "~/.m2/repository/io/delta/"], stream_output=True)
+    # No need to rm -rf local .ivy and .m2 io/delta caches.
+    # They were deleted in `prepare`, and then new artifacts were locally published to them
```
Comment (tdas) [inline python/run-tests.py:125]: this comment does not make sense in the context of the current code. its like saying "i did not steal a bank" out of the blue in a conversation, it raises more questions and confuses people. :)

File: python/run-tests.py [hunk 7]
```diff
@@ -100,6 +103,15 @@ def run_python_style_checks(root_dir):
 
 
 def run_pypi_packaging_tests(root_dir):
+    """
+    We want to test that the PyPi artifact for this delta version can be generated,
+    locally installed, and used in pyton tests.
+
+    We will uninstall any existing local delta artifact.
```
Comment (tdas) [inline python/run-tests.py:110]: this is not entirely true any more right ... we dont remove the maven artifact

File: python/run-tests.py (hunk 5 above)
Comment (scottsand-db) [reply python/run-tests.py:43]: we weren't doing it before. i added the print_cmd parameter to the run_cmd function.
since we weren't doing it before, saw no reason to do it now.

File: python/run-tests.py (hunk 6 above)
Comment (scottsand-db) [reply python/run-tests.py:125]: noted. good point.

File: python/run-tests.py (hunk 5 above)
Comment (tdas) [reply python/run-tests.py:43]: might as well do it. better for debugging.

Comment (scottsand-db) [reply python/run-tests.py:43]: sgtm

**Review by tdas — APPROVED** [review]
